File: Daily/279.perfect-squares.js
/*
 * @lc app=leetcode id=279 lang=javascript
 *
 * [279] Perfect Squares
 */

// @lc code=start

/** @type {number[]} */
const memo = [];

/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
    if (memo[n] !== undefined) return memo[n];
    if (Number.isInteger(Math.sqrt(n))) {
        memo[n] = 1;
        return 1;
    }

    let minNum = Infinity;

    for (let i = 1; ; i++) {
        const square = i ** 2;
        if (square < n) {
            minNum = Math.min(minNum, 1 + numSquares(n - square));
        } else {
            break;
        }
    }

    memo[n] = minNum;
    return minNum;
};
// @lc code=end

File: Daily/279.perfect-squares.cpp
/*
 * @lc app=leetcode id=279 lang=cpp
 *
 * [279] Perfect Squares
 */
#include <cmath>
#include <cstdint>
#include <algorithm>
// @lc code=start
class Solution
{
public:
    int numSquares(int n)
    {
        /*
        dp[i] the least number of perfect square numbers sum to i

        base case
        dp[0] = 0

        for i from 1 to n
            for j^2 from 1 to i
                dp[i] = min(dp[i], dp[i-j^2] + 1)
        */

        int dp[n+1];
        std::fill(dp, dp + n + 1, INT32_MAX);
        dp[0] = 0;

        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j * j <= i; j++)
            {
                dp[i] = std::min(dp[i], dp[i - j * j] + 1);
            }
        }

        return dp[n];
    }
};
// @lc code=end
